hashicorp/terraform-provider-hcp
# Hce 343 hvn project

### :hammer_and_wrench: Description

This code includes changes to allow associated project ID to be optionally read from either the provider config or hvn directly.

Jira 🎟️ :

[Add project_id to HVN](https://hashicorp.atlassian.net/jira/software/c/projects/HCE/boards/550?selectedIssue=HCE-343)

[Add validator that checks for project in resource and provider config](https://hashicorp.atlassian.net/jira/software/c/projects/HCE/boards/550?selectedIssue=HCE-342)

🧪 : Test Locally

To test locally, I executed the following scenarios. They can be reproduced using the steps below. Between each test scenario, delete the `terraform.lock.hcl` file and run `make dev` to recompile any terraform changes. The code below would be part of each of the example terraform configurations.
```
terraform {
	required_providers {
		hcp = {
			source  = "localhost/providers/hcp"
			version = "0.0.1"
		}
	}
}
```

1. Configure both the provider and hvn with the `project_id` fields set to distinct, existing projects. Run `terraform plan `to confirm the project id assigned is the id set in the hvn.
```
provider "hcp" {
	project_id = "valid-project-UUID" 
}

resource "hcp_hvn" "test" {
	hvn_id            = "test-hvn"
        project_id        = "valid-project-UUID" 
	cloud_provider    = "aws"
	region            = "us-west-2"
}
```
3.  Configure the hvn only with a valid project id. Run `terraform plan` to confirm the project id assigned is the id set in the resource.
```
provider "hcp" {}
resource "hcp_hvn" "test" {
	hvn_id            = "test-hvn"
        project_id        = "valid-project-UUID" 
	cloud_provider    = "aws"
	region            = "us-west-2"
}
```
5. Configure the project only with a valid project id. Run `terraform apply`. Run `terraform destroy`. Prior to responding 'yes' to confirm destroying the resource, confirm that the listed project id matches the id specified in the provider.
```
provider "hcp" {
	project_id = "valid-project-UUID" 
}

resource "hcp_hvn" "test" {
	hvn_id            = "test-hvn"
	cloud_provider    = "aws"
	region            = "us-west-2"
}
```
7. Configure either the provider or hvn with a project_id not corresponding to a valid UUID format (`ex: id-random-345` ). Run `terraform plan`. Confirm the error specifies a valid UUID is required.
```
provider "hcp" {
	project_id = "invalid-project-UUID" 
}

resource "hcp_hvn" "test" {
	hvn_id            = "test-hvn"
	cloud_provider    = "aws"
	region            = "us-west-2"
}
```
9. Configure neither the provider nor hvn with a project id. Confirm that you are able to successfully run `terraform apply` and that the hvn created has a project ID corresponding to the oldest existing project on your organization. Every organization per the current implementation auto-generates a project id on creation.
```
provider "hcp" {}

resource "hcp_hvn" "test" {
	hvn_id            = "test-hvn"
         project_id       = "valid-project-UUID" 
	cloud_provider    = "aws"
	region            = "us-west-2"
}
```

(https://hashicorp.atlassian.net/jira/software/c/projects/HCE/boards/550?selectedIssue=HCE-342)
### :building_construction: Acceptance tests

- [ ] Are there any feature flags that are required to use this functionality?
- [ ] Have you added an acceptance test for the functionality being added?
- [ ] Have you run the acceptance tests on this branch?

Output from acceptance testing:


```
$ make testacc TESTARGS='-run=TestAccAwsHvnOnly'
==> Checking that code complies with gofmt requirements...
golangci-lint run --config ./golangci-config.yml 
TF_ACC=1 go test ./internal/... -v -run=TestAccAwsHvnOnly -timeout 210m -parallel=10
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/clients	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/input	(cached) [no tests to run]
=== RUN   TestAccAwsHvnOnly
--- PASS: TestAccAwsHvnOnly (135.43s)
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider	135.951s
...
```
